### PR TITLE
Create Event in case of OOBM failure

### PR DESCRIPTION
**Pull request by GabrielBrascher**
## Description

Out of band power operations are constantly issued to assess the host power status. However, if such actions fail the Admins receive a warning message on the log. In order to help Admins to easily track such failures, this PR creates an Event each time an Out-of-band power operation fails.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

File: server/src/main/java/org/apache/cloudstack/outofbandmanagement/PowerOperationTask.java [hunk 1]
```diff
@@ -43,8 +53,26 @@ public void run() {
         try {
             service.executePowerOperation(host, powerOperation, null);
         } catch (Exception e) {
-            LOG.warn(String.format("Out-of-band management background task operation=%s for host id=%d failed with: %s",
-                    powerOperation.name(), host.getId(), e.getMessage()));
+            LOG.warn(String.format("Out-of-band management background task operation=%s for host id=%d name=%s failed with: %s",
+                    powerOperation.name(), host.getId(), host.getName(), e.getMessage()));
```
Comment (wido) [inline server/src/main/java/org/apache/cloudstack/outofbandmanagement/PowerOperationTask.java:57]: In general in CloudStack we try to not log the 'hostid' as it doesn't say anything. The ID can only be fetched from the DB and not from the API and thus UI.

File: server/src/main/java/org/apache/cloudstack/outofbandmanagement/PowerOperationTask.java [hunk 2]
```diff
@@ -43,8 +53,26 @@ public void run() {
...
+
+            if(CREATE_EVENT_ON_OOBM_FAILURE.value()) {
+                String eventMessage = String.format("Error while issuing out-of-band management action (%s) for host (id: %d, name: %s)",
+                        powerOperation.name(), host.getId(), host.getName());
```
Comment (wido) [inline server/src/main/java/org/apache/cloudstack/outofbandmanagement/PowerOperationTask.java:61]: Same applies here

File: server/src/main/java/org/apache/cloudstack/outofbandmanagement/PowerOperationTask.java [hunk 3]
```diff
@@ -17,16 +17,26 @@
...
 
-public class PowerOperationTask implements Runnable {
+public class PowerOperationTask implements Runnable, Configurable {
     public static final Logger LOG = Logger.getLogger(PowerOperationTask.class);
 
     final private OutOfBandManagementService service;
     final private Host host;
     final private OutOfBandManagement.PowerOperation powerOperation;
 
+    public static final ConfigKey<Boolean> CREATE_EVENT_ON_OOBM_FAILURE = new ConfigKey<Boolean>("Advanced", Boolean.class, "create.event.on.oob.failure", "true",
```
Comment (wido) [inline server/src/main/java/org/apache/cloudstack/outofbandmanagement/PowerOperationTask.java:36]: I know there is no naming scheme here, but isn't it easier if OOB related config keys are prefixed with *oob*?

<pre>oob.create.event.on.failure</pre>

For example

File: server/src/main/java/org/apache/cloudstack/outofbandmanagement/PowerOperationTask.java (hunk 1 above)
Comment (GabrielBrascher) [reply server/src/main/java/org/apache/cloudstack/outofbandmanagement/PowerOperationTask.java:51]: Makes sense, I am going to keep only the host name then.

File: server/src/main/java/org/apache/cloudstack/outofbandmanagement/PowerOperationTask.java (hunk 3 above)
Comment (GabrielBrascher) [reply server/src/main/java/org/apache/cloudstack/outofbandmanagement/PowerOperationTask.java]: Thanks for the suggestion, @wido.
I am removing that global settings config key. There are too many config keys already, I am leaving to add a new one only if it is really necessary.

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by GabrielBrascher** [comment]
@DaanHoogland I took the liberty to aim this one to 4.15.
Please let me know if this one is getting too late then I move it to 4.16 ;)

**Comment by DaanHoogland** [comment]
no tests added but for regression,
@blueorangutan test

**Review by rhtyd — APPROVED** [review]
LGTM - did not test it

**Review by wido — APPROVED** [review]
LGTM

Admins should be notified if CloudStack can't reach the OOB of a node. As this will be used not very often it's good that CloudStack keeps testing the reachability of the OOB so that when it's needed it also works.

**Comment by GabrielBrascher** [comment]
@rhtyd @DaanHoogland @wido @PaulAngus here follows a screenshot with the warning events:
image

**Comment by DaanHoogland** [comment]
@PaulAngus can we put this in for 4.15?

**Comment by DaanHoogland** [comment]
@PaulAngus very small and meets all criteria, merging

**Pull request merged**